File: src/chain-observers/collectors/jsonrpc/rc-account-collector.ts
import Long from "long";
import { RcAccountClassifier } from "../../classifiers";
import { IRcAccount } from "../../classifiers/rc-account-classifier";
import { DataEvaluationContext } from "../../factories/data-evaluation-context";
import { CollectorBase, TAvailableClassifiers } from "../collector-base";

const MAX_RC_ACCOUNT_GET_LIMIT = 1000;

export class RcAccountCollector extends CollectorBase<RcAccountClassifier> {
  private readonly rcAccounts: Record<string, number> = {};

  protected pushOptions(data: RcAccountClassifier["optionsType"]): void {
    this.rcAccounts[data.rcAccount] = (this.rcAccounts[data.rcAccount] || 0) + 1;
  }

  protected popOptions(data: RcAccountClassifier["optionsType"]): void {
    this.rcAccounts[data.rcAccount] = (this.rcAccounts[data.rcAccount] || 1) - 1;

    if (this.rcAccounts[data.rcAccount] === 0)
      delete this.rcAccounts[data.rcAccount];
  }

  public async get(data: DataEvaluationContext) {
    const rcAccounts: Record<string, IRcAccount> = {};

    const accountNames = Object.keys(this.rcAccounts);
    for (let i = 0; i < accountNames.length; i += MAX_RC_ACCOUNT_GET_LIMIT) {
      const chunk = accountNames.slice(i, i + MAX_RC_ACCOUNT_GET_LIMIT);

      const startFindRcAccounts = Date.now();
      const { rc_accounts: apiRcAccounts } = await this.worker.chain!.api.rc_api.find_rc_accounts({ accounts: chunk });
      data.addTiming("rc_api.find_rc_accounts", Date.now() - startFindRcAccounts);

      for(const rcAccount of apiRcAccounts)
        rcAccounts[rcAccount.account] = {
          name: rcAccount.account,
          rcManabar: {
            currentMana: Long.fromValue(rcAccount.rc_manabar.current_mana),
            max: Long.fromValue(rcAccount.max_rc),
            lastUpdateTime: new Date(rcAccount.rc_manabar.last_update_time * 1000)
          }
        };
    }

    return {
      /*
       * Instruct TypeScript typings that RcAccountClassifier.name is actualy a Classifier name we expect.
       * This is required for the bundlers to properly deduce the type of the classifier in data evaluation context.
       */
      [RcAccountClassifier.name as "RcAccountClassifier"]: {
        rcAccounts
      } satisfies TAvailableClassifiers["RcAccountClassifier"]
    };
  };
}

File: python/workerbee/chain_observers/collectors/jsonrpc/rc_account_collector.py
"""RcAccountCollector — fetches RC data via rc_api.find_rc_accounts."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ...classifiers.rc_account_classifier import RcAccountClassifier
from ..set_managed_collector import SetManagedCollector

if TYPE_CHECKING:
    from ...factories.data_evaluation_context import TCollectorEvaluationContext


class RcAccountCollector(SetManagedCollector):
    _options_key = "rc_account"

    async def get(self, data: TCollectorEvaluationContext) -> dict[str, Any]:
        if not self._tracked:
            return {RcAccountClassifier.__name__: {"rc_accounts": {}}}

        rc_accounts: dict[str, Any] = {}
        for chunk in self._tracked_chunks():
            with data.add_timing("rc_api.find_rc_accounts"):
                result = await self.worker.chain.api.rc_api.find_rc_accounts(accounts=chunk)

            for acc in result.rc_accounts:
                name = acc.account
                rc_accounts[name] = {
                    "name": name,
                    # ``max_rc`` lives on the RcAccount, not inside rc_manabar, so fold it
                    # into the manabar dict where ManabarCollector reads it (TS IRcAccount).
                    "rc_manabar": {
                        "current_mana": acc.rc_manabar.current_mana,
                        "max_rc": acc.max_rc,
                        "last_update_time": acc.rc_manabar.last_update_time,
                    },
                }

        return {RcAccountClassifier.__name__: {"rc_accounts": rc_accounts}}
